File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys
from collections import deque


# nxm 보드(맵), 0 : 빈칸(이동가능) 1 : 벽(이동불가)
# (1,1) -> (n,m) 이동. 최단 경로로 이동하기 -> 다익스트라.
# 시작하는 칸, 끝나는 칸 포함해서 센다.
# 이동하는 도중 한개의 벽을 부수고 이동하는 것이 좀 더 경로가 짧아진다면, 벽을 한 개까지 부수고 이동.
# 벽을 부순횟수 1과 부순횟수0은 다른 상태 -> 상태관리가 필요.

n,m = map(int, input().split())
g = [list(map(int, input())) for _ in range(n)]
INF = int(1e9)
# dis[x][y][z]에서 (x,y)까지 도달하는 데 필요한 최단거리를 구하는데 z는 벽을 부순 횟수를 나타냄.
# z = 0 벽 안 부숨, z = 1 벽 부숨.
# 따라서 다른 상태. 각각의 상태에서 최단거리 계산해야 한다.
dis = [[[INF] * 2 for _ in range(m)] for _ in range(n)]
startX,startY = 0,0
endX,endY = n-1,m-1
dx = [-1,0,1,0]
dy = [0,1,0,-1]
def isInner(x,y):
    if 0<=x<n and 0<=y<m:
        return True
    return False
def BFS(startX,startY):
    global res
    dq = deque()
    dq.append((startX,startY,0,1)) # 시작 위치(x,y), 벽을 부순 횟수, 현재까지의 경로를 저장 (시작위치도 포함)

    while dq:
        x,y,cnt,distance = dq.popleft() # 현재 경로의 위치, 벽을 부순 횟수(부쉈는지 안 부쉈는지), 현재 경로의 길이
        if (x,y) == (endX,endY): # 도달
            res = min(res, distance)
            continue # 도착점 도달 시 더 할 필요 없음
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not isInner(nx,ny): continue
            if g[nx][ny] == 0: # 빈칸 -> 이동 가능
                if distance + 1 < dis[nx][ny][cnt]: # 현재 경로보다 작은 경우에만 갱신
                    dis[nx][ny][cnt] = distance + 1
                    dq.append((nx,ny,cnt,distance + 1))
            else: # 벽 -> 벽을 부술 수 있는지
                if cnt == 0 and distance + 1 < dis[nx][ny][1]:
                    dis[nx][ny][1] = distance + 1 # 벽을 부숴서 거리 갱신
                    dq.append((nx,ny,1,distance + 1)) # 벽 부순 것으로 갱신 후 현재 경로 진행.

res = INF
BFS(startX,startY)
if res == INF:
    print(-1)
else:
    print(res)

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.*;
import java.util.*;


public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int n,m, endX,endY, min_dis;
    static int[][] g; // ! 그래프
    static int[][][] dis;
    static int INF = 10000000;

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
    private static class Point{
        int x,y,dis,cnt; // ! 현재 좌표, 거리, 벽을 부쉈는지 체크

        Point(int x, int y, int dis, int cnt) {
            this.x = x;
            this.y = y;
            this.dis = dis;
            this.cnt = cnt;
        }
    }

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        g = new int[n][m];
        dis = new int[n][m][2]; // ! 벽을 안부쉈음 0 벽을 부쉈음 1
        // ! 즉 dis[x][y][0] , dis[x][y][1] 벽을 부쉈을 때의 최단거리, 아닐 때의 최단거리 등등. 계산 가능.
        endX = n-1;
        endY = m-1;
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < m; j++) {
                g[i][j] = line.charAt(j) - '0'; // 인트로 변환
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                Arrays.fill(dis[i][j], INF);
            }
        }

        min_dis = INF;
        BFS(0, 0);
        if (min_dis == INF) {
            System.out.println(-1);
        } else {
            System.out.println(min_dis);
        }

    }

    public static boolean isInner(int x, int y) {
        if (x >= 0 && x < n && y >= 0 && y < m) {
            return true;
        }
        return false;
    }

    public static void BFS(int startX, int startY) {
        Deque<Point> dq = new ArrayDeque<>();
        dq.offer(new Point(startX, startY, 1, 0)); // ! 시작 좌표 포함
        dis[startX][startY][0] = 1; // ! 시작좌표 세팅.

        while (!dq.isEmpty()) {
            Point now = dq.poll();
            if (now.x == endX && now.y == endY) {
                if (now.dis < min_dis) {
                    min_dis = now.dis;
                    break; // ! 도착점에서 다시 할 필요 없음.
                }
            }

            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if(!isInner(nx,ny)) continue; // ! 좌표 내부이면 continue

                if (g[nx][ny] == 0) { // ! 이동 가능
                    if (now.dis + 1 < dis[nx][ny][now.cnt]) {
                        dis[nx][ny][now.cnt] = now.dis + 1;
                        dq.offer(new Point(nx, ny, now.dis + 1, now.cnt));
                    }
                } else if (g[nx][ny] == 1) { // ! 벽
                    if (now.cnt == 0) { // ! 벽을 부순 적이 없다면.
                        if (now.dis + 1 < dis[nx][ny][1]) { // ! 벽을 부쉈을 때보다 작으면
                            dis[nx][ny][1] = now.dis + 1;
                            dq.offer(new Point(nx, ny, now.dis + 1, 1));
                        }
                    }
                }
            }
        }
    }

    public static void 보드출력(int[][] g) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print(g[i][j] + " ");
            }
            System.out.println();
        }
    }

}